Fixing 503 Service Unavailable errors during fetch phase

jainankitk opened the pull request:
When ESRejectedExecutionException gets thrown on the coordinating node while trying to fetch hits, the resulting exception will hold no shard failures, hence `503` is used as the response status code. In that case, `429` should be returned instead. Also, the status code should be taken from the cause if available whenever there are no shard failures instead of blindly returning `503` like we currently do.

jainankitk committed:
Fixing 503 Service Unavailable errors during fetch phase

jainankitk committed:
Addressing checkstyle failure

elasticmachine commented:
Pinging @elastic/es-search

javanna commented:
hi @jainankitk thanks for your contribution! Would you mind signing our CLA please?

jainankitk commented:
Signed the CLA, thanks!

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java:
```diff
@@ -85,8 +85,9 @@ private static Throwable deduplicateCause(Throwable cause, ShardSearchFailure[]
-            return RestStatus.SERVICE_UNAVAILABLE;
+            // if no successful shards, the failure could have happened even before the fetch phase started
+            // so try to get the status from the cause instead of returning SERVICE_UNAVAILABLE blindly
+            return ExceptionsHelper.status(this.getCause());
```
the only doubt I have about this change is that we may end up returning 500 instead of 503, for instance whenever the cause is null, and probably in other cases too. Should we at least keep 503 in case the cause is null?

jainankitk replied:
Leaving null to 500 was intentional. Ideally speaking, being not able to parse the exception correctly is server side (elasticsearch here) error, and should be 500. Receiving 500 on client side indicates potential bug and helps us iterate faster to address other scenarios.

javanna replied:
I get what you are saying but here we are returning 503 for a specific case, where there are no shared failures yet we got an error. I am not sure we should go from 503 to 500 in that case, but I will discuss it with the team.

jainankitk replied:
Do you have any updates on this?

javanna replied:
not yet, thanks for the ping!

jimczi replied:
we discussed with @javanna and we think that we can preserve the old behavior if the cause is null. In such case we can continue to return a `503`, this is important for actions like https://github.com/elastic/elasticsearch/blob/9e706966282eaadafc77727059dd4194cbed492a/server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java#L152 that implies that setting shard failures to an empty array will return the right status.
So can you change this to:
````
return getCause() == null ? RestStatus.SERVICE_UNAVAILABLE : ExceptionsHelper.status(this.getCause());
````

jainankitk replied:
@jimczi Thanks for highlighting this use case. I have updated the PR with proposed changes.

javanna reviewed (commented):
Thanks for your contribution. The fix looks correct to me, I left a question though on potentially returning 500 instead of 503 in some cases. I am wondering why you mentioned a potential race between `onShardFailure` and `onPhaseFailure`, it seems to me that this would reproduce all the time whenever you get rejections on the fetch phase. Did you experience something different? Thanks again.

jainankitk commented:
> it seems to me that this would reproduce all the time whenever you get rejections on the fetch phase. Did you experience something different?
Yes, error will always reproduce for rejections on the fetch phase. I mentioned potential race as that is underlying cause, in case there are scenarios other than fetch phase where this can happen.

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java:
```diff
@@ -85,8 +85,9 @@ private static Throwable deduplicateCause(Throwable cause, ShardSearchFailure[]
-            // if no successful shards, it means no active shards, so just return SERVICE_UNAVAILABLE
-            return RestStatus.SERVICE_UNAVAILABLE;
+            // if no successful shards, the failure could have happened even before the fetch phase started
+            // so try to get the status from the cause instead of returning SERVICE_UNAVAILABLE blindly
```
shall we explicitly mention rejection from the fetch phase given that this is the problem, and it does not have to do with timing?

jainankitk replied:
@javanna I have updated the comment to explicitly mention rejected exception

jimczi reviewed (changes requested):
Thanks @jainankitk, I left one comment regarding the old behavior but I agree that we should try to extract the cause if it exists. I also wonder if the original issue should be avoided by forcing the execution of the fetch phase in the coordinating node. We already force the execution of the search phase here https://github.com/elastic/elasticsearch/blob/9e706966282eaadafc77727059dd4194cbed492a/server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java#L176 so I wonder if we should do the same for the fetch search phase. I'll comment on the issue but I don't think we should have a rejection in the coordinating node that is not coming directly from a shard.

jainankitk committed:
Addressing review comments

jainankitk committed:
Updating the code comment to explicitly mention EsRejectedExecutionException for fetch phase

jainankitk commented:
Thanks @javanna and @jimczi for the review. Have updated the branch with proposed changes.
> so I wonder if we should do the same for the fetch search phase. I'll comment on the issue but I don't think we should have a rejection in the coordinating node that is not coming directly from a shard.

Yes, I did wonder the same. Similarly, I find [query validation happening on non-coordinator nodes](https://github.com/elastic/elasticsearch/issues/26799) kind of confusing.

jimczi reviewed (approved):
Thanks @jainankitk . The change looks good to me.

javanna commented:
test this please

javanna commented on server/src/test/java/org/elasticsearch/action/search/SearchPhaseExecutionExceptionTests.java:
```diff
@@ -121,4 +123,33 @@ public void testToAndFromXContent() throws IOException {
+            new EsRejectedExecutionException("ES rejected execution of fetch phase"), shardSearchFailures);
+
+        assertNotEquals(actual.status(), RestStatus.TOO_MANY_REQUESTS);
+    }
```
can we add a test for the case where we return 503 as we have empty failures and null cause?

jainankitk replied:
Good point @javanna. Have added another test.

javanna commented on server/src/test/java/org/elasticsearch/action/search/SearchPhaseExecutionExceptionTests.java:
```diff
@@ -121,4 +123,33 @@ public void testToAndFromXContent() throws IOException {
+        SearchPhaseExecutionException actual = new SearchPhaseExecutionException(phase, "unexpected failures",
+            new EsRejectedExecutionException("ES rejected execution of fetch phase"), shardSearchFailures);
+
+        assertNotEquals(actual.status(), RestStatus.TOO_MANY_REQUESTS);
```
could we make this assertion more accurate? Maybe use less failure types, so it is easier to assert on the returned status?

jainankitk replied:
Ok, have removed NullPointer and TimestampParsing exceptions. The return type for both Parsing and InvalidIndexTemplate exceptions is BAD_REQUEST.

javanna reviewed (changes requested):
I left two comments around testing, LGTM otherwise, thanks again!

jainankitk committed:
Adding unit test for empty shard list and null cause

jainankitk commented:
Have addressed the review comments. Excerpt from updated unit test run:
```
Suite: org.elasticsearch.action.search.SearchPhaseExecutionExceptionTests
  1> [2019-03-06T11:43:06,534][WARN ][o.e.b.JNANatives         ] [[SUITE-SearchPhaseExecutionExceptionTests-seed#[3CBFC0BC0D6A7C6D]]] Unable to lock JVM Memory: error=78, reason=Function not implemented
  1> [2019-03-06T11:43:06,538][WARN ][o.e.b.JNANatives         ] [[SUITE-SearchPhaseExecutionExceptionTests-seed#[3CBFC0BC0D6A7C6D]]] This can result in part of the JVM being swapped out.
  1> [2019-03-06T13:43:08,318][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testPhaseFailureWithoutSearchShardFailure] before test
  1> [2019-03-06T13:43:09,073][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testPhaseFailureWithoutSearchShardFailure] after test
  1> [2019-03-06T13:43:09,101][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testToAndFromXContent] before test
  1> [2019-03-06T13:43:09,420][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testToAndFromXContent] after test
  1> [2019-03-06T13:43:09,432][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testPhaseFailureWithSearchShardFailure] before test
  1> [2019-03-06T13:43:09,433][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testPhaseFailureWithSearchShardFailure] after test
  1> [2019-03-06T13:43:09,441][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testToXContent] before test
  1> [2019-03-06T13:43:09,453][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testToXContent] after test
  1> [2019-03-06T13:43:09,458][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testPhaseFailureWithoutSearchShardFailureAndCause] before test
  1> [2019-03-06T13:43:09,458][INFO ][o.e.a.s.SearchPhaseExecutionExceptionTests] [testPhaseFailureWithoutSearchShardFailureAndCause] after test
Completed [1/1] in 4.17s, 5 tests


> Task :server:unitTest
==> Test Summary: 1 suite, 5 tests
   [junit4] JVM J0:     1.10 ..     6.55 =     5.45s
   [junit4] Execution time total: 6.58 sec.
   [junit4] Tests summary: 1 suite, 5 tests
```

javanna reviewed (approved):
LGTM thanks @jainankitk I will merge this shortly

javanna commented:
retest this please

javanna commented:
heya @jainankitk may I ask you to merge master in please? There are some unrelated test failures that would be addressed that way. Thanks!

jainankitk committed:
Merge remote-tracking branch 'upstream/master'

jainankitk commented:
I ran the unit tests again. They look flaky to me, after 2-3 runs, I could finish most of the server tests, still stuck at:

```
* What went wrong:
Execution failed for task ':x-pack:plugin:ml:compileTestJava'.
> Could not resolve all files for configuration ':x-pack:plugin:ml:testCompileClasspath'.
   > Could not download ini4j.jar (org.ini4j:ini4j:0.5.2)
      > Could not get resource 'https://jcenter.bintray.com/org/ini4j/ini4j/0.5.2/ini4j-0.5.2.jar'.
         > Could not GET 'https://jcenter.bintray.com/org/ini4j/ini4j/0.5.2/ini4j-0.5.2.jar'.
            > Connection reset
```

javanna commented:
retest this please

javanna commented:
thanks @jainankitk !
